x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/ConcurrentPriorityQueue.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.transform.transforms;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.SortedSet;
+import java.util.TreeSet;
+import java.util.function.Function;
+
+/**
+ * {@link ConcurrentPriorityQueue} class provides a priority queue functionality with the additional capability of quickly finding a queue
+ * element by the element's id.
+ *
+ * The implementation of this queue is thread-safe and utilizes locks.
+ *
+ * @param <E> Type of the element
+ */
+public class ConcurrentPriorityQueue<E> implements Iterable<E> {
```

Comment:
It's best to avoid the name `ConcurrentPriorityQueue` as it sounds like a generic class that can be reused and this class does not implement `Queue`. You could go for something super specific like `TransformTaskOrderedQueue`

Reply:
> out of curiosity: did you considered the priority queue from lucene (its abstract, so provides a base implementation)?

You mean `org.apache.lucene.util.PriorityQueue`?
It is kind of related to @davidkyle question regarding using the additional map.
If we want to lookup by transform id (**not** the priority) quickly, then we need the map and, therefore, we need a custom priority queue class. If not, we can use the lucene queue or even the java `TreeSet` directly.
BTW, how many transforms can there be on a node at maximum?

> It's best to avoid the name ConcurrentPriorityQueue as it sounds like a generic class that can be reused and this class does not implement Queue. You could go for something super specific like TransformTaskOrderedQueue

Done (chosen name: `TransformScheduledTaskQueue`).

Reply:
> BTW, how many transforms can there be on a node at maximum?

Afaik there is no hard limit, at least for transforms. It seems p-tasks aren't limited either. The only limit I can think of is threadpools, but as we are using the generic threadpool I think even that isn't limited.